Lib/idlelib/pyshell.py

```diff
@@ -678,9 +675,18 @@ def runsource(self, source):
         filename = self.stuffsource(source)
         # at the moment, InteractiveInterpreter expects str
         assert isinstance(source, str)
-        # InteractiveInterpreter.runsource() calls its runcode() method,
-        # which is overridden (see below)
-        return InteractiveInterpreter.runsource(self, source, filename)
+        #if isinstance(source, str):
+        #    from idlelib import iomenu
+        #    try:
+        #        source = source.encode(iomenu.encoding)
+        #    except UnicodeError:
+        #        self.tkconsole.resetoutput()
+        #        self.write("Unsupported characters in input\n")
+        #        return
+        # II.runsource() calls .runcode(), overridden below.
+        with warnings.catch_warnings():
```

Comment:
This is great!

Just remember to remove the commented out code above before merging...

Reply:
When I created the PR on 2019 Aug 16, runsource comprised stuffsource(), a comment and assert, the inactivated code, and a try-finally block that wapped  the override comment II.runsource() and  and that fiddled with warnings in 'finally'.

The comment and assert were added and the code inactivated by MvL in 2007.  The commit message was "Expect that source strings are unicode".  Since str was bytes at the time, the message, comment, and assert do not make much sense to me.  In any case, the comment and assert can be deleted.  If IDLE were to pass bytes, there would be an error anyway.

I removed the inactivated code on Sept 16 in #16198 for bpo-38183 as an unrelated change.  (A separate PR would have been better.)  I only left it in the merge resolution because I was not sure then it should go (it should).

The try-finally was from KBK in 2001 and 2004.  On Aug 26, in #15500 for this same issue (bpo-37824), I redid shell warnings and removed 'try' and the finally block, leaving the comment and II.runsource().  At sometime, I wrapped the call in a warnings context in a failed attempt to prevent the triple warning (which is why I did not push it before).  The wrapper should go unless tests show that it is needed.

Reply:
It appears that II.runsource is only use for code input after '>>>'.  For instance, 'python -m idlelib -c "0 is "' only prints the warning once even without the filter.  The patch otherwise moves it from the console to Shell.  The current file level warnings manipulation does not save and restore filters.  So I will leave the filter here until such time as code or codeop are patched (see issue) to make it unneeded in IDLE.